### ISSUE TITLE
Metrics "framework" for JIT

### ISSUE THREAD

**Issue by mikedn** [dotnet/coreclr#18330]
While hacking the JIT quite often I end up with a simple question - "how many time does X happen?":
* How many JCC instructions are generated?
* How many side effects trees are extracted?
* How many range checks are encountered/removed?

So I end up adding variables, `printf`s and what not. And it's always a quick improvisation, done to be used once and then thrown away.

I think it would be useful to have some basic metrics framework in the JIT that would let people to simply add a line like:
```C++
metrics.Increment("ExtractedSideEffect");
```
pretty much anywhere in the JIT code and have such counters dumped to the console like mem stats are. Or included in disassembly comments so that they show up in diffs and maybe be used by jit-analyze.

Counters should be collected per method and if possible per phase as well. And of course, some basic aggregations should be included as well.

The API should be as easy to use as possible - the above example shows the use of a string as a counter identifier (so I guess the implementation would be using these strings as a hashtable key). This would avoid the need to add a new enum member - such an enum typically resides in a header that when changed triggers the compilation of all JIT files.  Of course, an enum based overload could also provided, that would be useful for counters that are useful long term.

@dotnet/jit-contrib Thoughts?

**Comment by AndyAyersMS** [comment]
I think this is worth pursuing, yes. I have done similar throwaway things to track various internal event frequencies.

And with PMI/SPMI we could start cross indexing method instances with these metrics, so one could for example find the set of methods with some jit-relevant property by query.

**Comment by BruceForstall** [comment]
I like the idea. We should make sure it's easy to enable for Release, but not in Release by default.

**Comment by BruceForstall** [comment]
Maybe the first measure of success would be how many of the existing off-by-default "metrics" could be converted to a standard format? E.g., all the defines in jit.h starting with `CALL_ARG_STATS`.

**Comment by mikedn** [comment]
Some example numbers I got in the past by using throw away instrumentation are in this range check post: https://github.com/dotnet/coreclr/issues/12639#issuecomment-335909456

In addition to counters (e.g. Removed range checks) there are also values where the maximum value is reported (e.g. Max search depth) and timings (e.g. Map def time). We gather phase timings but if you ae interested in what's going with just a part of a phase then you need to roll your own timer.

Speaking of phases - it's probably worth linking to #16360. The current phase infrastructure needs some improvement.

**Comment by BruceForstall** [comment]
> Which brings the question if it makes sense to attempt to handle all these stats through a common path.

If possible, yes, but I would be judicious: there's no point forcing all possible metrics into one model unless that makes sense.

**Comment by mikedn** [comment]
A rough sketch: https://gist.github.com/mikedn/5502f45f427f60b89bee03de42a04b1d

Basically you define "metrics":
```c++
Metric<SumValue> emitInstructionCount("test.count");
Metric<MinValue, MaxValue, AvgValue> emitTest("test.min-max-avg");
```
and `Compiler` holds a metric "store" that provides storage for metrics and the means to update them:
```c++
c.m_metrics.Increment(emitInstructionCount);
c.m_metrics.Add(emitTest, 42);
```
Might be a bit too complicated, at least due to the use of `std::tuple` which I'm not sure I can re-create in `jitstd`.

The example code prints something like:
```
Method A
test.count: 2
test.min-max-avg:
        min: 2
        max: 100
        avg: 48

Method B
test.count: 1
test.min-max-avg:
        min: 1
        max: 200
        avg: 67

Total
test.count: 3
test.min-max-avg:
        min: 1
        max: 200
        avg: 57
```

**Comment by BruceForstall** [comment]
Doing it this way requires creating a Metric object for each type of metric you want, and make sure it is exposed to all possible users. Standard, but perhaps not as lightweight as I thought you were going for.

I thought you were considering a design were everything was by string, and no formal C++ objects would get created for an object. You would just do:

```
metrics.Increment("ExtractedSideEffect");

or maybe:

metrics.Add("ExtractedSideEffect", 1);
```
If the name wasn't previously seen, it could be magically created and zeroed. Presumably you could also optionally initialize them and set their "type":
```
metrics.Init("ExtractedSideEffect", 0, Metrics::Count);
metrics.Init("min-max-avg", Metrics::MinMaxAvg);
metrics.Init("foo", Metrics::Histogram, "10,20,30,40,100,200,1000"); // specify histogram buckets
```
maybe the "Init" wouldn't even be required to run "first", if it was only to specify display info.

Something like this wouldn't give you compile-time checking, but also would be trivial to add to any single source file or set of source files, without worrying about putting declarations in header files, etc.

**Comment by mikedn** [comment]
> I thought you were considering a design were everything was by string, and no formal C++ objects would get created for an object. You would just do:

Yep, I was hoping to do something like that but I can't figure out a way to do that without using a hashtable. And despite the fact that this code is not going to be used in normal release builds I'd still try to keep its impact as low a possible so it can be used to measure execution time as well.

> but also would be trivial to add to any single source file or set of source files, without worrying about putting declarations in header files, etc.

It's still possible to avoid adding stuff to headers in most cases - for example you can just add a few metric definitions to rangecheck.cpp and use them in that file. Header declarations may be needed in cases where a JIT "component" is spread throughout multiple .cpp files, like in emitter's case.

Ultimately it's a matter of trade-offs and any feedback is highly useful.

**Comment by AndyAyersMS** [comment]
I think I'd go for something pretty simple here. For instance a declarative table syntax (like many other things we have in the jit) that creates a metric ID enum and a backing static table of fixed per-metric info (like name, hierarchy, etc). Then the counts vector is dynamic, indexed by ID.

This doesn't directly support min/max/histogram type stuff but that can be supported by keeping other "global" count vectors and aggregating into them at appropriate points. Min/Max/Average would be easy and there are also clever ways to estimate percentiles without keeping all observations around.

**Comment by mikedn** [comment]
> For instance a declarative table syntax (like many other things we have in the jit) that creates a metric ID enum and a backing static table of fixed per-metric info (like name, hierarchy, etc).

Yes, an enum is an option. The main reason I avoided that is that we'll end up with all metrics declared in a header file that when changed triggers recompilation of the entire JIT. Not a big deal, just a minor annoyance.

**Comment by AndyAyersMS** [comment]
There are ways around this must-be-in-header requirement on windows -- with linker section pragmas and such, you can build a self-assembling list -- but I'm not sure it is worth the trouble, or would port well to other build setups.
